packages/frontend/src/contexts/unirep.js

```diff
@@ -0,0 +1,144 @@
+import { makeAutoObservable } from 'mobx'
+import { SERVER } from '../config'
+
+export default class Unirep {
+  allSignUps = []
+  signUpsByUserId = {}
+  signUpsByAttesterId = {}
+  allAttestations = []
+  attestationsByAttesterId = {}
+  totalPosRep = 0
+  totalNegRep = 0
+  totalRep = this.totalPosRep - this.totalNegRep
+  allEpochs = []
+  attesterIds = []
+  currentAttesterStats = []
+
+  constructor(state) {
+    makeAutoObservable(this)
+    this.state = state
+    if (typeof window !== 'undefined') {
+      //   this.load()
+    }
+  }
+
+  //   async load() {}
+
+  async loadAllSignUps() {
+    const url = new URL(`api/unirep/signups`, SERVER)
+    const data = await fetch(url.toString()).then((r) => r.json())
+    this.allSignUps = data
+    this.ingestSignUps(data)
+  }
+
+  async ingestSignUps(signups) {
+    this.signUpsByUserId = new Map()
+    this.signUpsByAttesterId = new Map()
+    for (let i = 0; i < signups.length; i++) {
+      let userId = signups[i].commitment
+      let attesterId = signups[i].attesterId
+      if (this.signUpsByUserId.has(userId)) {
+        this.signUpsByUserId.get(userId).push(signups[i])
+      } else {
+        this.signUpsByUserId.set(userId, [signups[i]])
+      }
+
+      if (this.signUpsByAttesterId.has(attesterId)) {
+        this.signUpsByAttesterId.get(attesterId).push(signups[i])
+      } else {
+        this.signUpsByAttesterId.set(attesterId, [signups[i]])
+      }
+    }
+    console.log('userSignups:', this.signUpsByUserId)
+    console.log('attesterSignUps:', this.signUpsByAttesterId)
+    console.log(
+      'one users signups with map.get:',
+      this.signUpsByUserId.get(
+        '20665588589671275613306006678457435446746460926274194220015986402668071993482'
+      )
+    )
+    console.log(
+      'one users signups by id:',
+      this.signUpsByUserId[
+        '20665588589671275613306006678457435446746460926274194220015986402668071993482'
+      ]
+    )
+  }
```

Comment:
atm I'm duplicating at lot of data; still working out which approach will be the most practical, but would like to get some input on what I've got so far.

- I need to load ALL signups, attestations, and epochs to inform the `HomePage`, but I only need the lists that are returned, not the mappings. I created the mappings thinking I would then access them from `AttesterPage` and `UserPage` and potentially doing away with `User` and `Attester` contexts. But I was not having success accessing those `..ByUserId` and .`.ByAttesterId` mappings so I switched back to developing the `Attester` context.

- I'm currently stuck on how to access the data inside the mappings. I can log out the Maps (here and from my page components); they look how I expect and contain all the data I want but I'm unable to display the values. This is my console from the above, and I get the same when I try to get inside the Map from a component, map[id] is always undefined. Can you see what I am doing wrong?

<img width="1215" alt="Screen Shot 2023-01-24 at 1 58 18 PM" src="https://user-images.githubusercontent.com/79813968/214412080-faa5f5c7-ede9-4b2c-acc2-8be4eb32dff2.png">

- I have a lot of the data I need from the database, but I'm missing everything to do with timestamps and I don't see how to get them. The only table in the db that has a timestamp field is `Attestation`, and each attestation entry shows `"timestamp": "0"`. Is the `_id` field in each entry (all tables except `Epoch` the event id? and am I supposed to use that to find the timestamp for the event? So I don't currently have
   - userSignUp and attestation timestamps
   - attester deployment time (I know this is not an event currently emitted)
   - epoch duration / transition time
   - hashchain processing time

Reply:
> I'm currently stuck on how to access the data inside the mappings. I can log out the Maps (here and from my page components); they look how I expect and contain all the data I want but I'm unable to display the values. This is my console from the above, and I get the same when I try to get inside the Map from a component, map[id] is always undefined

Make sure to access values in a `Map` using the `.get` property. Accessing using `map[val]` will return undefined because values aren't set as keys on the object. In the example on line 60 above it's printing undefined. Not sure if that is the problem or not, I'd need to see an instance in the component to be sure.

> I have a lot of the data I need from the database, but I'm missing everything to do with timestamps and I don't see how to get them

Accessing timestamps for events is a bit of a pain. You'll need to load the block information for the block number the event was emitted. This is probably best done in the synchronizer when processing events. So when an event is seen the block data is loaded before/during processing and the timestamp is extracted and put in the database.

In the synchronizer there is a function [`processEvents`](https://github.com/Unirep/explorer/blob/routes/packages/backend/src/singletons/Synchronizer.mjs#L209). This function accepts an array of events and handles calling the relevant function for each specific event. I think this would be a good place to put the timestamp loading logic. You can think of this as a few distinct steps:
- Get the list of unique block numbers
- Check if there is already a timestamp stored in the database for each block number
- For any block number that doesn't already have a timestamp load the block data and save the timestamp

You'll need another database collection that looks like

```js
{
  name: 'BlockTimestamp',
  primaryKey: 'number',
  rows: [
    ['number', 'Int'],
    ['timestamp', 'Int'],
  ],
}
```

If you do this in `processEvents` then each event handler is guaranteed to have local access to the timestamp for the block it was emitted in. Then you can put the timestamp directly on the object to avoid having to load it later.

Let me know if that makes sense

Reply:
maybe : )
my last commit [1e53ba2](https://github.com/Unirep/explorer/pull/4/commits/1e53ba2b8314a682129daed9d6ca4740f1df0f55) is changes to the synchronizer. I feel like I understood the way you explained it, but I can't see the effects of what I've done.

<img width="1075" alt="Screen Shot 2023-01-29 at 7 30 59 PM" src="https://user-images.githubusercontent.com/79813968/215375123-53f552be-faa0-4b0a-b005-e8ef75d6f786.png">


I added this to `schema.js` (inside the @unirep/core package) and it was added to my db..
<img width="1075" alt="Screen Shot 2023-01-29 at 7 34 18 PM" src="https://user-images.githubusercontent.com/79813968/215375496-ab2dbd87-eefa-4953-9940-0ff0a03c1017.png">

..but I also added the `timestamp` field to the other tables and they did not show up.

I'm guessing that here (`index.mjs`) would be the place to make these changes to the schema? but atm I'm not getting how that could be done.
<img width="1075" alt="Screen Shot 2023-01-29 at 7 37 28 PM" src="https://user-images.githubusercontent.com/79813968/215375791-ce3ab6e8-fd92-41f5-ac97-7049cee20a03.png">

<hr>

> Make sure to access values in a `Map` using the `.get` property. Accessing using `map[val]` will return undefined because values aren't set as keys on the object.

Ok, thanks, having much better progress now!
but having a lot of errors: 'someMap.get is not a function'... still investigating: its not consistent, sometimes I'm able to get and render from the mappings and sometimes it errors.

So.. I had thought I was supposed to be able to access `Map` values with brackets bc isn't that what you're doing here in `CanonSection.jsx`?

<img width="896" alt="Screen Shot 2023-01-29 at 7 57 26 PM" src="https://user-images.githubusercontent.com/79813968/215378135-8147549a-9a4d-4fce-b84d-705a135de5e6.png">
and also this example you made up for me earlier

<img width="810" alt="Screen Shot 2023-01-29 at 8 05 43 PM" src="https://user-images.githubusercontent.com/79813968/215378949-178bd59e-663a-4bc2-9b42-938ed789d7f4.png">

how are these different from what I'm trying to do?

Reply:
> my last commit [1e53ba2](https://github.com/Unirep/explorer/pull/4/commits/1e53ba2b8314a682129daed9d6ca4740f1df0f55) is changes to the synchronizer. I feel like I understood the way you explained it, but I can't see the effects of what I've done.

Check [here](https://github.com/Unirep/explorer/pull/4/commits/1e53ba2b8314a682129daed9d6ca4740f1df0f55#r1090489266)

> I'm guessing that here (index.mjs) would be the place to make these changes to the schema? but atm I'm not getting how that could be done.

Yes, schema management is a bit hacky right now. In index.mjs you can do

```js
_schema.push({
  name: 'BlockTimestamp',
  primaryKey: 'number',
  rows: [
    ['number', 'Int'],
    ['timestamp', 'Int'],
  ],
})
```

> So.. I had thought I was supposed to be able to access Map values with brackets bc isn't that what you're doing here in CanonSection.jsx?

No, I don't use a `Map` object anywhere, I use simple objects (like `{}`) and just put properties on them. I imagine the inconsistent errors are coming from mixing `Map` and raw object. Definitely choose one or the other and use them everywhere or it will become a nightmare.

[This](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map#objects_vs._maps) is a great comparison of maps vs objects. Generally maps are preferable, especially in production grade applications - i just set a bad example in my prototype projects.